docs/proposals/formatted-output.md

```diff
@@ -0,0 +1,359 @@
+
+- `reference`: full artifact reference by digest, e.g, `$REGISTRY/$REPO@$DIGEST`
+- `referenceByTags`: array, pushed tags by reference, e.g. `$REGISTRY/$REPO@TAG1`
+- `mediaType`: media type of the pushed file (layer) 
+- `digest`: digest of the pushed file (layer) 
+- `size`: file size in bytes
+- `annotations`: contains arbitrary metadata for the image manifest
+
+For example, push a file and two tags to a repository and show the descriptor of the image manifest in pretty JSON format. 
+
+```bash
+oras push $REGISTRY/$REPO:$TAG1, $TAG2 sbom.spdx vul-scan.json --format json 
+```
+
+```json
+{
+  "reference": "localhost:5000/oras@sha256:4a5b8c83d153f52afdfcb422db56c2349aae3bd5ecf8338a58353b5eb6681c45",
+  "mediaType": "application/vnd.oci.image.manifest.v1+json",
+  "digest": "sha256:4a5b8c83d153f52afdfcb422db56c2349aae3bd5ecf8338a58353b5eb6681c45",
+  "size": 820,
+  "annotations": {
+    "org.opencontainers.image.created": "2023-12-15T09:41:54Z"
+  },
+  "artifactType": "json/example",
+  "referenceByTags": [
+    "$REGISTRY/$REPO:$TAG1",
+    "$REGISTRY/$REPO:$TAG2"
+  ]
+}
+```
+
+> [!NOTE]
+> When pushing a folder to filesystem, the output fields are the same as the example of pushing files as above.
+
+Push a folder to a repository and filter out the value of `reference` and `artifactType` of the pushed artifact in the standard output.
+
+```bash
+oras push $REGISTRY/$REPO:$TAG sample-folder --format '{{.reference}}, {{.mediaType}}'
+```
+
+```console
+localhost:5000@sha256:85438e6598bf35057962fff34399a362d469ca30a317939427fca6b7a289e70d, application/vnd.oci.image.manifest.v1+json
+```
+
+### oras attach
+
+When using `oras attach` with the flag `--format`, the following fields should be formatted into JSON output:
+
+- `reference`: full reference by digest of the referrer file
+- `mediaType`: media type of the referrer
+- `size`: referrer file size in bytes
+- `digest`: digest of the attached referrer file
+- `artifactType`: artifact type of the referrer
+- `annotations`: contains arbitrary metadata in a referrer
+
+For example, attach two files to an image and show the descriptor metadata of the referrer in JSON format.
+
+```bash
+oras attach $REGISTRY/$REPO:$TAG --artifact-type example/report-and-sbom vul-report.json:example/vul-scan sbom.spdx:example/sbom --format json
+```
+
+```json
+{
+  "reference": "localhost:5000/oras@sha256:0afd0f0c35f98dcb607de0051be7ebefd942eef1e3a6d26eefd1b2d80f2affbe",
+  "mediaType": "application/vnd.oci.image.manifest.v1+json",
+  "digest": "sha256:0afd0f0c35f98dcb607de0051be7ebefd942eef1e3a6d26eefd1b2d80f2affbe",
+  "size": 923,
+  "annotations": {
+    "org.opencontainers.image.created": "2023-12-15T08:59:21Z"
+  }
+  "artifactType": "doc/example"
+}
+```
+
+### oras discover
+
+View an artifact's referrers. The default output should be listed in a tree view.
+
+```bash
+oras discover $REGISTRY/$REPO:$TAG --format tree
```

Comment:
What about `--verbose` for tree output?

Reply:
Currently, the output shows `annotations` of each referrer in the tree view as below. In theory, tree view is intended to display the artifact reference relationship graph, i.e. **demonstrate how referrers associate with the root artifact**. It looks a bit confusing and complicated that the annotations data also showed in the tree view. If users want to see `annotations` of each referrer, they could use JSON view. 

Current behavior:

```
oras discover localhost:5000/my-ubuntu:latest -v
localhost:5000/my-ubuntu@sha256:a3785f78ab8547ae2710c89e627783cfa7ee7824d3468cae6835c9f4eae23ff7
├── application/vnd.cncf.notary.signature
│   └── sha256:8dee8cb9a1334595545e3baf15c3eeed13c4b35ae08e3ab32e1df31fb152dc1d
│       ├── io.cncf.notary.x509chain.thumbprint#S256: '["79e91aa1e109a16df87d200e493fd3d33c67253f76d41334d7f7c29c00ba55b3"]'
│       └── org.opencontainers.image.created: "2024-04-02T10:32:55Z"
└── sbom/example
    └── sha256:50fd0dc107d84b5e7b402688000a7ed3aaf8a2692d5cb74da5277fa3c4cecf15
        └── org.opencontainers.image.created: "2024-05-10T07:57:10Z"
```

So it would be much neat if we keep the verbose output consistent with the default output in the tree view of `oras discover`:

```
localhost:5000/my-ubuntu@sha256:a3785f78ab8547ae2710c89e627783cfa7ee7824d3468cae6835c9f4eae23ff7
├── application/vnd.cncf.notary.signature
│   └── sha256:8dee8cb9a1334595545e3baf15c3eeed13c4b35ae08e3ab32e1df31fb152dc1d
└── sbom/example
    └── sha256:50fd0dc107d84b5e7b402688000a7ed3aaf8a2692d5cb74da5277fa3c4cecf15
```

Reply:
I suggest that we design the tree view holistically in the next iteration.

Reply:
SGTM. Linking this comment to the verbose log enhancement issue https://github.com/oras-project/oras/issues/1091#issuecomment-2105503928